### PR TITLE
feat: implement SHPlonk VerifyOpeningProof

### PR DESCRIPTION
**Pull request by Insun35**
# Description

Implemented SHPlonk multi-opening proof verifying logic.
See https://eprint.iacr.org/2020/081.pdf

### Refactoring
- Since the `transcripts` have been moved from zk to crypto namespace, 
   it doesn't need to return the value of proofs.
   It can write proofs directly into a transcript.

- Generalized `Poly` to `PolyOracle` for common use in `CreateOpeningProof()` and `VeriryOpeningProof()`.
  `CreateOpeningProof()` uses the `Poly` type, and `VerifyOpeningProof()` uses the `Commitment` type for `PolyOracle`.

- Changed SHPlonk template parameters from Point type to Curve type to use Pairing.

### Fix
- Replaced `absl::flat_hash_map` to `std::vector` since the order of `poly_openings` affects linear combinations.

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
@Insun35 Can you fix commit message of 46916ec? VeriryOpeningProof -> VerifyOpeningProof.

File: tachyon/crypto/commitments/kzg/shplonk.h [hunk 1]
```diff
@@ -178,22 +187,173 @@ class SHPlonk : public UnivariatePolynomialCommitmentScheme<
...
+    const absl::btree_set<PointDeepRef>& super_point_set =
+        grouper.super_point_set();
+
+    Field first_z_diff_inverse = Field::Zero();
+    Field first_z = Field::Zero();
+
+    std::vector<G1JacobianPointTy> l_commitments;
+    l_commitments.reserve(grouped_poly_openings_vec.size());
+    size_t i = 0;
+    for (const auto& [poly_openings_vec, points_ref] :
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/shplonk.h:225]: How about
```suggestion
    for (const auto& [poly_openings_vec, points] :
```
or rename member name in `GroupedPolynomialOpenings`?

File: tachyon/crypto/commitments/kzg/shplonk.h [hunk 2]
```diff
@@ -178,22 +187,173 @@ class SHPlonk : public UnivariatePolynomialCommitmentScheme<
...
-          typename _Commitment>
-struct VectorCommitmentSchemeTraits<
-    SHPlonk<G1PointTy, G2PointTy, MaxDegree, _Commitment>> {
+template <typename CurveTy, size_t MaxDegree, typename _Commitment>
+struct VectorCommitmentSchemeTraits<SHPlonk<CurveTy, MaxDegree, _Commitment>> {
  public:
   constexpr static size_t kMaxSize = MaxDegree + 1;
   constexpr static bool kIsTransparent = false;
 
+  // using G1PointTy = typename CurveTy::G1Curve::G1AffinePoint;
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/shplonk.h:355]: Please remove this.

File: tachyon/zk/plonk/halo2/prover_test.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 #include "gtest/gtest.h"
 
+#include "tachyon/math/elliptic_curves/bn/bn254/bn254.h"
 #include "tachyon/math/elliptic_curves/bn/bn254/g1.h"
 #include "tachyon/math/elliptic_curves/bn/bn254/g2.h"
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/prover_test.h:11]: Do we still need this?

File: tachyon/zk/plonk/halo2/BUILD.bazel [hunk 4]
```diff
@@ -121,6 +121,7 @@ tachyon_cc_library(
     hdrs = ["prover_test.h"],
     deps = [
         ":prover",
+        "//tachyon/math/elliptic_curves/bn/bn254",
         "//tachyon/math/elliptic_curves/bn/bn254:g1",
         "//tachyon/math/elliptic_curves/bn/bn254:g2",
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/BUILD.bazel:126]: Do we still need this?

File: tachyon/crypto/commitments/kzg/shplonk_unittest.cc [hunk 5]
```diff
@@ -40,52 +46,75 @@ class SHPlonkTest : public testing::Test {
...
-  std::vector<PolynomialOpening<Poly>> poly_openings_;
+  std::vector<Commitment> commitments_;
+  std::vector<PolynomialOpening<Poly>> prover_openings_;
+  std::vector<PolynomialOpening<Poly, Commitment>> verifier_openings_;
   SimpleTranscriptWriter<Commitment> writer_;
 };
 
 }  // namespace
 
 TEST_F(SHPlonkTest, CreateProof) {
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/shplonk_unittest.cc:111]:
```suggestion
TEST_F(SHPlonkTest, CreateAndVerifyProof) {
```

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/crypto/commitments/polynomial_openings.h [hunk 6]
```diff
@@ -150,17 +150,15 @@ class PolynomialOpeningGrouper {
...
 
     // Group |poly_openings_grouped_by_poly| by points.
     // [x₀, x₁, x₂]: [P₀, P₁, P₂]
     // [x₂, x₃]: [P₃]
     // [x₄]: [P₄]
-    absl::flat_hash_map<absl::btree_set<PointDeepRef>,
-                        std::vector<PolyOracleDeepRef>>
-        poly_openings_grouped_by_poly_and_points =
-            GroupByPoints(poly_openings_grouped_by_poly);
+    std::vector<PointGroupedPair> poly_openings_grouped_by_poly_and_points =
```
[the remark this answers is not in the document]
Comment (Insun35) [reply tachyon/crypto/commitments/polynomial_openings.h:160]: Prover and Verifier must construct each opening set from different input types (`Poly` and `Commitment`).
If we use a Btree map, the order may be different cuz the keys are different on each Prover and Verifier side. So the values may be different in linear combination. 
For `super_point_set` where the order doesn't matter, I used Btree map.

File: tachyon/crypto/commitments/kzg/shplonk.h (hunk 1 above)
Comment (Insun35) [reply tachyon/crypto/commitments/kzg/shplonk.h:225]: Agreed. I will rename the member name to `point_refs`

File: tachyon/zk/plonk/halo2/prover_test.h (hunk 3 above)
Comment (Insun35) [reply tachyon/zk/plonk/halo2/prover_test.h:11]: You're right, we don't need them. I'll remove them.

File: tachyon/crypto/commitments/polynomial_openings.h [hunk 7]
```diff
@@ -16,54 +16,54 @@
...
 // A single polynomial with a single opening.
-template <typename Poly>
+template <typename Poly, typename PolyOracle = Poly>
 struct PolynomialOpening {
   using Field = typename Poly::Field;
   using Point = typename Poly::Point;
 
-  // P₀
-  base::DeepRef<const Poly> poly;
+  // polynomial -> P₀ || commitment -> C₀
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/polynomial_openings.h:24]: Polynomial P0 or Commitment C0. ("Arrow" and "||" are ambiguous for me).

File: tachyon/crypto/commitments/polynomial_openings.h [hunk 8]
```diff
@@ -16,54 +16,54 @@
 namespace tachyon::crypto {
 
 // A single polynomial with a single opening.
-template <typename Poly>
+template <typename Poly, typename PolyOracle = Poly>
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/polynomial_openings.h:19]: Please add comment related to the following context.
```
`CreateOpeningProof()` uses `Poly` type, and `VerifyOpeningProof()` uses
`Commitment` type for `PolyOracle`.
```

File: tachyon/crypto/commitments/polynomial_openings.h [hunk 9]
```diff
@@ -16,54 +16,54 @@
...
-  PolynomialOpenings(base::DeepRef<const Poly> poly,
+  PolynomialOpenings(base::DeepRef<const PolyOracle> poly_oracle,
                      std::vector<Field>&& openings)
-      : poly(poly), openings(std::move(openings)) {}
+      : poly_oracle(poly_oracle), openings(std::move(openings)) {}
 };
 
 // Multi polynomials with multi openings grouped by shared points.
-template <typename Poly>
+template <typename Poly, typename PolyOracle = Poly>
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/polynomial_openings.h:54]: Please modify the comment to encompass both cases when `PolyOracle` is 'Poly' and when it's 'Commitment'.

File: tachyon/crypto/commitments/polynomial_openings.h [hunk 10]
```diff
@@ -237,7 +270,8 @@ class PolynomialOpeningGrouper {
       PolyOracleDeepRef poly_oracle, PointDeepRef point) {
     auto it = std::find_if(
         poly_openings.begin(), poly_openings.end(),
-        [poly_oracle, point](const PolynomialOpening<Poly, PolyOracle>& poly_opening) {
+        [poly_oracle,
+         point](const PolynomialOpening<Poly, PolyOracle>& poly_opening) {
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/polynomial_openings.h:274]: It is weird. It seems like the formatter is malfunctioning.

File: tachyon/crypto/commitments/kzg/shplonk.h [hunk 11]
```diff
@@ -178,22 +187,172 @@ class SHPlonk : public UnivariatePolynomialCommitmentScheme<
...
+
+    Field first_z_diff_inverse = Field::Zero();
+    Field first_z = Field::Zero();
+
+    std::vector<G1JacobianPointTy> l_commitments;
+    l_commitments.reserve(grouped_poly_openings_vec.size());
+    size_t i = 0;
+    for (const auto& [poly_openings_vec, point_refs] :
+         grouped_poly_openings_vec) {
+      // commitments_0 = [C₀, C₁, C₂]
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/kzg/shplonk.h:227]: Please use the |variable| symbol for variable names.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: tachyon/crypto/commitments/polynomial_openings.h (hunk 10 above)
Comment (Insun35) [reply tachyon/crypto/commitments/polynomial_openings.h:274]: Formatting was not done in the previous commit. I'll rebase it.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**